### PR TITLE
S3 blob store ultimate rebase

### PR DESCRIPTION
**Pull request by Arsnael**
Just going wild and trying stuff... Don't mind it too much :)

### CODE REVIEW

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
Downgrade to zenko/cloudserver 8.1.20
* 1 green build
* 2 builds failing on the flaky tests
* 2 builds failing for non related issues
* 1 failing on deployment tests

Conclusion: Still the same, does not solve anything, need to look somewhere else

**Comment by Arsnael** [comment]
Attempt to reuse @blackheaven 's work with a pool (as suggested from @chibenwa )

A bit more work to do if adopted but I would like to send a first batch of builds with this just to see

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java [hunk 1]
```diff
@@ -99,12 +103,21 @@
             .region(region.asAws())
             .serviceConfiguration(pathStyleAccess)
             .build();
+
+        clientPool = PoolBuilder.from(Mono.fromCallable(clientCreator))
+            .acquisitionScheduler(Schedulers.elastic())
+            .destroyHandler(client -> Mono.fromRunnable(client::close))
+            .maxPendingAcquireUnbounded()
+            .sizeBetween(4, 64)
```
Comment (chibenwa) [inline server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java:111]: If possible let's have an unbounded pool, and make the size configurable with ridiculously high defaults otherwize.

Comment (Arsnael) [reply server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java:111]: Yes I saw your comments on the related PR

There was a comment regarding the warmup, to not do it in constructor but Startable pattern fashion too

**Comment by Arsnael** [comment]
5 builds:
* 2 green builds
* 2 failed build but for missing deps during compilation (so BS not related error probably caused by network issues)
* 2 failed deployment tests

No flaky tests anymore and the ratio seems acceptable. I think we can go forward with this.

I will work on it to answer @chibenwa 's comments

**Comment by Arsnael** [comment]
I just did :
* rebase on latest master
* squash the fixups
* reworked the git history
* did small fixes on the way (license formatting, change the number of the ADR to match our current state, ...)

Ready for review :)

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            .endpointOverride(configuration.getEndpoint())
+            .region(region.asAws())
+            .serviceConfiguration(pathStyleAccess)
+            .build();
+
+        clientPool = PoolBuilder.from(Mono.fromCallable(clientCreator))
+            .acquisitionScheduler(Schedulers.elastic())
+            .destroyHandler(client -> Mono.fromRunnable(client::close))
+            .maxPendingAcquireUnbounded()
+            .sizeUnbounded()
```
Comment (chibenwa) [inline server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java:112]: Wonderfull !

File: src/site/markdown/server/install/guice-cassandra-rabbitmq-swift.md [hunk 3]
```diff
@@ -50,7 +50,7 @@ You need to have a Cassandra, ElasticSearch and RabbitMQ instance running. You c
 $ docker run -d -p 9042:9042 --name=cassandra cassandra:3.11.3
 $ docker run -d -p 9200:9200 --name=elasticsearch --env 'discovery.type=single-node' docker.elastic.co/elasticsearch/elasticsearch:6.3.2
 $ docker run -d -p 5672:5672 -p 15672:15672 --name=rabbitmq rabbitmq:3.8.1-management
-$ docker run -d -p 5000:5000 -p 8080:8080 -p 35357:35357 --name=swift linagora/openstack-keystone-swift:pike
+$ docker run -d -p 8080:8000 --name=s3 scality/s3server:6018536a
```
Comment (chibenwa) [inline src/site/markdown/server/install/guice-cassandra-rabbitmq-swift.md:53]: Additional env variables needed?

File: src/site/xdoc/server/config-blobstore.xml [hunk 4]
```diff
@@ -118,99 +102,14 @@
...
-                    </subsection>
-                </subsection>
                 <subsection name="ObjectStorage AWS S3 Configuration">
                     <dl>
                         <dt><strong>objectstorage.s3.endPoint</strong></dt>
                         <dd>S3 service endpoint</dd>
 
+                        <dt><strong>objectstorage.s3.region</strong></dt>
+                        <dd>S3 region</dd>
+
```
Comment (chibenwa) [inline src/site/xdoc/server/config-blobstore.xml:112]: Don't forget to update the new Antora documentation as well!

File: upgrade-instructions.md [hunk 5]
```diff
@@ -19,6 +19,19 @@ Change list:
...
+
+### Swift support has been dropped
+
+Date ??/07/2020
+
+JIRA: https://issues.apache.org/jira/browse/JAMES-3028
+
+Concerned product: Distributed James
+
+OpenStack's Swift support has been dropped, please use Cassandra or AWS S3 instead.
```
Comment (chibenwa) [inline upgrade-instructions.md:32]: `OpenStack's Swift APIs support has been dropped, please use Cassandra or AWS S3 instead. Please note that modern swift version supports S3 APIs out of the box, thus OpenStack Swift can still be used through the S3 API.`

File: dockerfiles/run/docker-compose.yml [hunk 6]
```diff
@@ -8,9 +8,10 @@ services:
       - cassandra
       - tika
       - rabbitmq
-      - swift
+      - s3
     entrypoint: bash -c "java -Dworking.directory=/root/ -Dlogback.configurationFile=/root/conf/logback.xml -jar james-server.jar"
-    image: linagora/james-rabbitmq-project:latest
+#    image: linagora/james-rabbitmq-project:latest
+    image: james_run
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply dockerfiles/run/docker-compose.yml:14]: I think that was part of a wip, needs to be changed back to the line above I believe

File: README.adoc [hunk 7]
```diff
@@ -226,9 +226,9 @@ You need a running *rabbitmq* in docker. To achieve this run:
 
     $ docker run -d --name=rabbitmq rabbitmq:3.8.1-management
 
-You need a running *swift* objectstorage in docker. To achieve this run:
+You need a running *s3* objectstorage in docker. To achieve this run:
 
-    $ docker run -d --name=swift linagora/openstack-keystone-swift:pike
+    $ docker run -d -p 8080:8000 --name=s3 scality/s3server:6018536a
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply README.adoc:231]: Will recheck that... it's not even scality/s3server we use but zenko/cloudserver as well

[the remark this answers is not in the document]
Comment (chibenwa) [reply README.adoc:231]: Yes, let's advertise the same image than the one we use in the tests

File: dockerfiles/run/docker-compose.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply dockerfiles/run/docker-compose.yml:14]: I believe too

**Comment by Arsnael** [comment]
```
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] [ERROR] selectThenFetchWithExistingMessages  Time elapsed: 0.845 s  <<< ERROR!
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] org.apache.james.mpt.protocol.ProtocolSession$InvalidServerResponseException: 
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] 
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] Location: /org/apache/james/imap/scripts/SelectThenFetchWithExistingMessages.test:51
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] LastClientMsg: A06 DELETE selectTest
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] Expected: 'A06 OK DELETE completed\.'
[f4c9f74e35a10c9a1b8402be74ef7b7f0037c1c3] Actual   : 'A06 NO DELETE processing failed.'
```

**Comment by Arsnael** [comment]
@rouazana 

Ran the rebase S3 with the `JmapSendMessagesSimulation` against a zenko cloudserver : 

screenshot--2020 09 14-16_59_13

**Comment by chibenwa** [comment]
That's nice but what is it like against master <3

**Comment by Arsnael** [comment]
> That's nice but what is it like against master <3

yep yep running one atm :)

**Comment by chibenwa** [comment]
test this please (deployment tests)

File: docs/modules/servers/pages/distributed/run-docker.adoc [hunk 8]
```diff
@@ -51,9 +51,9 @@ You need a running *rabbitmq* in docker. To achieve this run:
 
     $ docker run -d --name=rabbitmq rabbitmq:3.8.1-management
 
-You need a running *swift* objectstorage in docker. To achieve this run:
+You need a running *Zenko Cloudserver* objectstorage in docker. To achieve this run:
 
-    $ docker run -d --name=swift linagora/openstack-keystone-swift:pike
+    $ docker run -d -p 8080:8000 --env 'S3BACKEND=mem' --env 'REMOTE_MANAGEMENT_DISABLE=1' --env 'SCALITY_ACCESS_KEY_ID=accessKey1' --env 'SCALITY_SECRET_ACCESS_KEY=secretKey1' --name=s3 zenko/cloudserver:8.2.6
```
Comment (remk) [inline docs/modules/servers/pages/distributed/run-docker.adoc:56]: is it desirable to propose the mem backend by default here ?

File: README.adoc [hunk 9]
```diff
@@ -226,9 +226,9 @@ You need a running *rabbitmq* in docker. To achieve this run:
 
     $ docker run -d --name=rabbitmq rabbitmq:3.8.1-management
 
-You need a running *s3* objectstorage in docker. To achieve this run:
+You need a running *s3* compatible objectstorage in docker. To achieve this run:
 
-    $ docker run -d -p 8080:8000 --name=s3 scality/s3server:6018536a
+    $ docker run -d -p 8080:8000 --env 'S3BACKEND=mem' --env 'REMOTE_MANAGEMENT_DISABLE=1' --env 'SCALITY_ACCESS_KEY_ID=accessKey1' --env 'SCALITY_SECRET_ACCESS_KEY=secretKey1' --name=s3 zenko/cloudserver:8.2.6
```
Comment (remk) [inline README.adoc:231]: is it desirable to propose the mem backend by default here ?

File: docs/modules/servers/pages/distributed/run.adoc [hunk 10]
```diff
@@ -53,15 +53,15 @@ following command:
...
+You need to have a Cassandra, ElasticSearch, S3 and RabbitMQ instance
 running. You can either install the servers or launch them via docker:
 
 [source,bash]
 ----
 $ docker run -d -p 9042:9042 --name=cassandra cassandra:3.11.3
 $ docker run -d -p 9200:9200 --name=elasticsearch --env 'discovery.type=single-node' docker.elastic.co/elasticsearch/elasticsearch:6.3.2
 $ docker run -d -p 5672:5672 -p 15672:15672 --name=rabbitmq rabbitmq:3.8.1-management
-$ docker run -d -p 5000:5000 -p 8080:8080 -p 35357:35357 --name=swift linagora/openstack-keystone-swift:pike
+$ docker run -d -p 8080:8000 --env 'S3BACKEND=mem' --env 'REMOTE_MANAGEMENT_DISABLE=1' --env 'SCALITY_ACCESS_KEY_ID=accessKey1' --env 'SCALITY_SECRET_ACCESS_KEY=secretKey1' --name=s3 zenko/cloudserver:8.2.6
```
Comment (remk) [inline docs/modules/servers/pages/distributed/run.adoc:64]: is it desirable to propose the mem backend by default here ?

File: src/site/markdown/server/install/guice-cassandra-rabbitmq-swift.md [hunk 11]
```diff
@@ -44,13 +44,13 @@ You also need to generate a keystore in your conf folder with the following comm
...
 
-You need to have a Cassandra, ElasticSearch and RabbitMQ instance running. You can either install the servers or launch them via docker:
+You need to have a Cassandra, ElasticSearch, S3 and RabbitMQ instance running. You can either install the servers or launch them via docker:
 
 ```bash
 $ docker run -d -p 9042:9042 --name=cassandra cassandra:3.11.3
 $ docker run -d -p 9200:9200 --name=elasticsearch --env 'discovery.type=single-node' docker.elastic.co/elasticsearch/elasticsearch:6.3.2
 $ docker run -d -p 5672:5672 -p 15672:15672 --name=rabbitmq rabbitmq:3.8.1-management
-$ docker run -d -p 8080:8000 --name=s3 scality/s3server:6018536a
+$ docker run -d -p 8080:8000 --env 'S3BACKEND=mem' --env 'REMOTE_MANAGEMENT_DISABLE=1' --env 'SCALITY_ACCESS_KEY_ID=accessKey1' --env 'SCALITY_SECRET_ACCESS_KEY=secretKey1' --name=s3 zenko/cloudserver:8.2.6
```
Comment (remk) [inline src/site/markdown/server/install/guice-cassandra-rabbitmq-swift.md:53]: is it desirable to propose the mem backend by default here ?

**Comment by Arsnael** [comment]
master branch with swift locally, same scenario :

screenshot--2020 09 15-09_08_45

I think we can agree the rebase is useful...

**Comment by Arsnael** [comment]
Now if the only remaining comments are 'removing the env variable for backend mem in the doc, can we agree we can fix that upon merge, as the green is green already and this is doc change, and if there is nothing else can people approve it? 

Thank you :)

**Comment by chibenwa** [comment]
> Now if the only remaining comments are 'removing the env variable for backend mem in the doc', can we agree we can fix that upon merge, as the build is green already and this is doc change, and if there is nothing else can people approve it?

Or you can fire a fix for the doc, that we don't await to be :green_apple: , that's about the same

**Comment by Arsnael** [comment]
Doc fix, build was :green_apple: before

**Comment by Arsnael** [comment]
S3 rebase test with this time a zenko cloudserver with a file backend:

screenshot--2020 09 15-11_19_13

Still much less errors and better response time overall. I think we are good

**Review by remk — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**